File: sorting/shell_sort.py
data=[45,84,77,83,55,49,91,64,91,5,37,31,70,38,51]
#define the gap by yourself
gap=[5,2,1]

def shell_sort(array,gap):
    n=len(array)
    #start gapped insertion sort
    for current_gap in gap:
        for i in range(current_gap,n):
            #temp: the insertion number
            temp=array[i]
            j=i
            #if the unsorted array is bigger than insertion number, then shift right 
            while j>=0 and j-current_gap>=0 and array[j-current_gap]>temp:
                array[j]=array[j-current_gap]
                j-=current_gap
            #insert the insertion number
            array[j]=temp
    return array

print(shell_sort(data,gap))

File: sorting/shellsort.c
#include <stdio.h>
#define SIZE 15

void shellSort(int *, int *);
void printArray(int *);

int main(){
	int data[SIZE]={45,84,77,83,55,49,91,64,91,5,37,31,70,38,51};
	//define gap by yourself
	int gap[3]={5,2,1};
	shellSort(data,gap);
	printArray(data);
	return 0;
}

void shellSort(int *array, int *gap){
	int i,j,k,current_gap,temp;
	for(i=0;i<3;i++){
		current_gap=gap[i];
		for(j=current_gap;j<SIZE;j++){
			temp=array[j];
			k=j;
			while(k>=0&&k-current_gap>=0&&array[k-current_gap]>temp){
				array[k]=array[k-current_gap];
				k-=current_gap;
			}
			array[k]=temp;
		}
	}
}

void swap(int *a, int *b){
	int temp;
	temp=*a;
	*a=*b;
	*b=temp;
}

void printArray(int *array){
	int i;
	for(i=0;i<SIZE;i++){
		printf("%d ",array[i]);
	}
}
